File: two-pointers/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()

        for i in range(0, len(nums)-1):
            if i > 0 and nums[i] == nums[i-1]:
                continue
            
            l, r = i+1, len(nums)-1
            while l < r:
                curr = nums[i] + nums[l] + nums[r]
                if curr == 0:
                    res.append([nums[i], nums[l], nums[r]])
                    l += 1
                    while nums[l] == nums[l-1] and l < r:
                        l += 1
                elif curr < 0:
                    l += 1
                else:
                    r -= 1
        return res

File: array/3sum.rb
# @param {Integer[]} nums
# @return {Integer[][]}
def three_sum(nums)
    nums.sort!
    res = []
    0.upto(nums.size-1) do |i|
        next if i > 0 and nums[i] == nums[i-1]
        l = i+1
        r = nums.size-1
        while l < r
            curr_sum = nums[i] + nums[l] + nums[r]
            puts curr_sum
            if curr_sum == 0
                res << [nums[i], nums[l], nums[r]]
                l += 1
                while nums[l] == nums[l-1] and l < r
                    l += 1
                end
            elsif curr_sum > 0
                r -= 1
            else
                l += 1
            end
        end
    end
    res
end